internal/clients/compute/group/group.go

```diff
@@ -0,0 +1,419 @@
+package group
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	sdkgo "github.com/ionos-cloud/sdk-go/v6"
+
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/apis/compute/v1alpha1"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients/compute"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/utils"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// APIClient is a wrapper around IONOS Service
+type APIClient struct {
+	*clients.IonosServices
+}
+
+// Client is a wrapper around IONOS Service Group methods
+type Client interface {
+	CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error)
+	GetGroupID(group *sdkgo.Group) (string, error)
+	GetGroup(ctx context.Context, groupID string) (sdkgo.Group, *sdkgo.APIResponse, error)
+	GetGroupMembers(ctx context.Context, groupID string) ([]string, *sdkgo.APIResponse, error)
+	GetGroupResourceShares(ctx context.Context, groupID string) ([]v1alpha1.ResourceShare, *sdkgo.APIResponse, error)
+	CreateGroup(ctx context.Context, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	UpdateGroup(ctx context.Context, groupID string, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	AddGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	RemoveGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	UpdateGroupMembers(ctx context.Context, groupID string, membersIn MembersUpdateOp) error
+	AddResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	RemoveResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateGroupResourceShares(ctx context.Context, groupID string, sharesIn SharesUpdateOp) error
+	DeleteGroup(ctx context.Context, groupID string) (*sdkgo.APIResponse, error)
+	GetAPIClient() *sdkgo.APIClient
+}
+
+// MembersUpdateOp groups memberIDs in sets depending on the operation in which they will be used
+type MembersUpdateOp struct {
+	Add, Remove sets.Set[string]
+}
+
+// SharesUpdateOp groups resource shares in sets depending on the operation in which they will be used
+type SharesUpdateOp struct {
+	Add, Update, Remove sets.Set[v1alpha1.ResourceShare]
+}
+
+// CheckDuplicateGroup based on groupName
+func (cp *APIClient) CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error) {
```

Comment:
I suggest this name change
`func (cp *APIClient) GetGroupByName(ctx context.Context, groupName string) (sdkgo.Group, error) {`
The return by value will avoid any side effect, please see my comment few lines down in the return statement

Reply:
I think we can just return the ID of the duplicate if we found any, since that's what we are only interested in anyway, see ([df32e25](https://github.com/ionos-cloud/crossplane-provider-ionoscloud/pull/140/commits/df32e25c674273b256ff319f4d070564728cbeeb))